### PR TITLE
Pin: Update 3DS to include new parameters

### LINKED ISSUE #4719: Update Pin gateway
Update Pin gateway to use documented parameters:

https://pinpayments.com/developers/api-reference/charges#post-charges

### PR DESCRIPTION
**Pull request by hudakh**
Pin: Update 3DS secure to include new parameters

We have found we cannot use 3DS secure with Pin as the parameters do not match those documented by Pin. This closes #4719. 

Test Summary
Local: 5476 tests, 77236 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
Unit: 44 tests, 145 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
Remote: 21 tests, 65 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed

### CODE REVIEW

File: lib/active_merchant/billing/gateways/pin.rb
Comment (AMHOL) [inline lib/active_merchant/billing/gateways/pin.rb:189]: @hudakh this will still need to support the old style of 3DS as implemented in https://github.com/activemerchant/active_merchant/pull/3848, I'd suggest doing something like this:

```ruby
def add_3ds(post, options)
  if options[:three_d_secure]
    post[:three_d_secure] = {}

    if options[:three_d_secure][:enabled]
      post[:three_d_secure][:enabled] = true
      post[:three_d_secure][:fallback_ok] = options[:three_d_secure][:fallback_ok] if options[:three_d_secure][:fallback_ok]
      post[:three_d_secure][:callback_url] = options[:three_d_secure][:callback_url] if options[:three_d_secure][:callback_url]
    else
      post[:three_d_secure][:version] = options[:three_d_secure][:version] if options[:three_d_secure][:version]
      post[:three_d_secure][:eci] = options[:three_d_secure][:eci] if options[:three_d_secure][:eci]
      post[:three_d_secure][:cavv] = options[:three_d_secure][:cavv] if options[:three_d_secure][:cavv]
      post[:three_d_secure][:transaction_id] = options[:three_d_secure][:ds_transaction_id] || options[:three_d_secure][:xid]
    end
  end
end
```

Comment (hudakh) [reply lib/active_merchant/billing/gateways/pin.rb]: Thanks @AMHOL. I've added a commit with your suggestion.

File: test/remote/gateways/remote_pin_test.rb
Comment (AMHOL) [inline test/remote/gateways/remote_pin_test.rb:24]: Would be good to retain the old test here and also add the new version, perhaps updating to `@additional_options_3ds_passthrough = {version, eci, etc...}` and `@additional_options_3ds = {enabled, falback_ok, etc...}`, updating `test_successful_authorize_and_capture_with_passthrough_3ds` to use `@additional_options_3ds_passthrough` and adding `test_successful_authorize_and_capture_with_3ds`

File: test/unit/gateways/pin_test.rb
Comment (AMHOL) [inline test/unit/gateways/pin_test.rb:349]: Same as above, it would be good to retain these tests and add an additional test for the new style 3DS params.

File: test/remote/gateways/remote_pin_test.rb
Comment (hudakh) [reply test/remote/gateways/remote_pin_test.rb]: I've updated the tests.

File: test/remote/gateways/remote_pin_test.rb
Comment (AMHOL) [inline test/remote/gateways/remote_pin_test.rb:37]: Additional `end` here is causing a syntax error when trying to run remote tests, they pass with this removed.
```suggestion
```

**Comment by AMHOL** [comment]
Nice work @hudakh, changes look good to me and Unit/Remote tests pass :+1: 

Still needs workflow approval and review from an AM maintainer.

**Comment by hudakh** [comment]
> Nice work @hudakh, changes look good to me and Unit/Remote tests pass +1
> 
> Still needs workflow approval and review from an AM maintainer.

Thanks @AMHOL. I've cleaned up the commits and PR description in preparation for further review. Not sure if I need to assign to anyone.

**Comment by AMHOL** [comment]
I haven't done in the past, normally someone just comes along and picks it up.

**Comment by hudakh** [comment]
@pjhyett, @lancecarlson, @brynary can this be marked as ready for review please? I am not sure how to proceed further to get this merged.

**Comment by madpilot** [comment]
Is there not somewhere else to do this? My last PR on this repo was 11 years ago. I'm not the best person to be approving anything.

**Comment by hudakh** [comment]
> Is there not somewhere else to do this? My last PR on this repo was 11 years ago. I'm not the best person to be approving anything.

No worries. Your name was the only name listed for Pin under the contributors list.. I thought perhaps that would give you some super powers I do not have.

**Comment by hudakh** [comment]
@jessiagee Can you please assist here? I see you were the reviewer of the last Pin-related merge request. I am not sure of next steps, but I feel this won't progress if I don't prod someone.

**Comment by raymzag** [comment]
following this too. trying to pin @rachelkirk as well who might have looked at previous Pin-related PRs before. 

Hey Rachel, would you be able to help or direct to someone who could help with progressing this PR?

**Comment by AMHOL** [comment]
@hudakh if this is blocking you, you can pull the gem from your fork using the following in your Gemfile:

```ruby
gem 'activemerchant', git: 'git@github.com:hudakh/active_merchant.git', branch: '1-update-pin-gateway'
# or
gem 'activemerchant', git: 'git@github.com:hudakh/active_merchant.git', ref: 'c716f13a9c18db7bb4550f175f42f981555af079'
```

Otherwise you'll have to wait for the next activemerchant release after this PR is merged.

**Comment by aenand** [comment]
Hello! Could you please elaborate on how you will need the change in AM (i.e. do you need a new release or just a merge to `master`)? Additionally can you provide more details on why you are making this change?

**Comment by AMHOL** [comment]
@aenand

@hudakh would need a merge/release in order to use this feature on the mainline.

The change updates the Pin gateway in order to add support for passing `three_d_secure` params to the `charges` endpoint as documented [here](https://pinpayments.com/developers/api-reference/charges#post-charges), whilst retaining support for the old style 3D Secure passthrough options in order to maintain support for existing integrations.

File: lib/active_merchant/billing/gateways/pin.rb
Comment (aenand) [inline lib/active_merchant/billing/gateways/pin.rb:87]: `verify` is a reserved action that attempts to verify if a payment method is ready for purchase. It is meant to be a low amount auth & void or 0 amount auth. This does not seem to be doing this so could we change the name of this method? Additionally I don't see this method being called anywhere, am I missing something?

File: lib/active_merchant/billing/gateways/pin.rb
Comment (aenand) [inline lib/active_merchant/billing/gateways/pin.rb:186]: There may be a situation where `options[:three_d_secure]` is not present so calling `options[:three_d_secure][:enabled]` would throw a `NoMethodError`. Could we add a guard at the top of this method to `return unless options[:three_d_secure]`?

File: lib/active_merchant/connection.rb
Comment (aenand) [inline lib/active_merchant/connection.rb:44]: Why drop the `.parse`?

**Review by aenand — CHANGES_REQUESTED** [review]

File: lib/active_merchant/billing/gateways/pin.rb
Comment (AMHOL) [reply lib/active_merchant/billing/gateways/pin.rb:186]: @aenand it's already guarded [here](https://github.com/activemerchant/active_merchant/pull/4720/files#diff-0e0df11ddf16e4e9c9d9225f2f82b44f460d63db9f13fe52b5324ff917f9248aR184).

Comment (aenand) [reply lib/active_merchant/billing/gateways/pin.rb:186]: Oops Github's UI hid it from me. thanks!

File: lib/active_merchant/connection.rb
Comment (hudakh) [reply lib/active_merchant/connection.rb:44]: > At some point, I was getting an error (as implied by the commit message). I am no longer getting errors so I will revert this change.

File: lib/active_merchant/billing/gateways/pin.rb
Comment (AMHOL) [reply lib/active_merchant/billing/gateways/pin.rb:87]: @hudakh perhaps we could go with `verify_3ds` or `verify_three_d_secure` here?

Comment (hudakh) [reply lib/active_merchant/billing/gateways/pin.rb]: @aenand, the method is called by the apps using the Pin gateway to verify that the charge was successful. I have renamed as per @AMHOL's suggestion.

**Comment by aenand** [comment]
There are some rubocop linter failures. Please run `bundle exec rubocop -a` to autocorrect them!

**Comment by hudakh** [comment]
> There are some rubocop linter failures. Please run `bundle exec rubocop -a` to autocorrect them!

I'm not familiar with running rubocop in different ruby versions, so I corrected the issues manually.

**Comment by AMHOL** [comment]
Pending review.

**Review by aenand — APPROVED** [review]
Sorry I lost this 🤦🏽 Looks good to me. Please squash the commits and add a changelog!

**Comment by AMHOL** [comment]
> Sorry I lost this 🤦🏽 Looks good to me. Please squash the commits and add a changelog!

@hudakh are you available to do this?

**Comment by AMHOL** [comment]
Done @aenand

**Comment by hudakh** [comment]
Awesome, thank you so much!

**Comment by AMHOL** [comment]
Rebased

**Comment by aenand** [comment]
@AMHOL i'm so sorry i missed this. it's saying there are merge conflicts, can you check those?

**Comment by AMHOL** [comment]
@aenand I am not seeing any merge conflict issues?

Screenshot 2024-02-06 at 10 45 39

**Comment by AMHOL** [comment]
@aenand is anything further needed to get this merged?

**Comment by AMHOL** [comment]
Still not stale.

**Comment by aenand** [comment]
Screenshot 2024-06-21 at 1 30 31 PM

Any idea why it's claiming merge conflicts?

**Comment by AMHOL** [comment]
Updated @aenand, it was the CHANGELOG entry.

**Pull request merged**
